main.go

```diff
@@ -43,33 +49,50 @@ func main() {
 		log.Fatal("All scrape endpoints are disabled. You need enable at least one endpoint.")
 	}
 
-	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
-		if r.URL.Path != "/" {
-			http.NotFound(w, r)
-			return
-		}
+	router := getHTTPRouter()
+
+	log.WithField("port", config.BindAddr).Info("Listening for scrapes.")
+	log.WithError(router.Run(config.BindAddr)).Fatal("Shutting down.")
+}
+
+func getHTTPRouter() *gin.Engine {
+	router := gin.Default()
+
+	rg := router.Group("/")
+	rg.GET("/liveness", func(ctx *gin.Context) {
 		log.WithFields(log.Fields{
-			"uri":    r.RequestURI,
-			"client": r.RemoteAddr,
-		}).Debug("Accessed Root endpoint")
-		http.Redirect(w, r, "/metrics", http.StatusMovedPermanently)
+			"uri":    ctx.Request.RequestURI,
+			"client": ctx.Request.RemoteAddr,
+		}).Debug("Accessed Liveness endpoint")
+		ctx.String(http.StatusNoContent, "")
 	})
-	http.HandleFunc("/liveness", func(w http.ResponseWriter, r *http.Request) {
+
+	if config.BasicAuth.Username != "" {
+		if config.BasicAuth.Password == "" {
+			log.Fatal("Must set basic-auth.password to enable basic auth.")
+		}
+
+		rg.Use(gin.BasicAuth(gin.Accounts{
+			config.BasicAuth.Username: config.BasicAuth.Password,
+		}))
```

Comment:
AFAIU, enabling BasicAuth middleware on the whole router group under `/` results that every endpoint is now under auth. Please don't do this for `/liveness` and `/` itself, only for `/metrics`. `/liveness` is needed by Kubernetes health checks and those shouldn't require auth. I see no reason why `/` (a simple redirect) and `/liveness` (health endpoint) should be behind auth. If my underlying assumption is wrong, please do correct me, I'm not that familiar with Gin :)

That said, it may be worth adding a simple http unit test that verifies that the liveness is not behind auth if basic auth is enabled if you have some spare time.